Modular Shield Generators, Also Adds a new perimeter algorithm to _helpers/math_dm

moocowswag opened the pull request:
What I have so far

https://www.youtube.com/watch?v=EXj-8yL6Y08&t



 Disclaimer im mostly making this pr for the experience if anything.
   Also anyone can replace the sprites if they want to I`m just making them so itl be done, idk what a plasteel machine is supposed to look like.

## About The Pull Request
https://hackmd.io/PV9h_VtgRleSR6FnAra1gw?both
Basically, this pr adds a new generator that generates a shield in the form of a ring.
There is a possibility for different shield shapes in the future.

The circuit boards to make these machines are unlocked by the advanced power manipulation tech

The entire shield

- Has one big healthpool (base of 50)
- Has a configurable radius
- Takes damage from hot gas and emp
- Has regeneration (it gets nerfed by alot of things that I have to talk about separately)
- When the shield is broken the generator must regenerate at 0.25x speed before coming online again
- can be configured to either only generate a shield externally (in space) or both internally and externally at the cost of half the regen rate
- Regen rate is also reduced based on the ratio of the shield radius and the max radius that the generator is capable of generating
- Alot more that im too tired to explain at the moment

The base stats of the shield (max health, regen rate, max radius) all scale with the tier of parts used in the main generator.

4 other machines (3 are finished code-wise as of now) the charger (regen rate), the relay (max radius) the well (max health) the node (allows for more connections) can all be connected to the main generator to boost its core stats based on their part tiers

In the future there is a possibility of specialization / modifications to the forcefield at the cost to the core stats
## Why It's Good For The Game

1. Engineers lack job content outside of making repairs to the station and getting bullied for letting the SM delam

2. We lack serious fortifications in this game that actually care about the number of people that are trying to breach them, take reinforced walls for instance, in an extreme scenario like a malf ai youll often have a crowd of people watching as one person disassembles the reinforced walls to get to the ai`s satellite.

The end goal is something that even if its seriously used people will be able to participate and have fun with breaching or building.

3. This pr lays a foundation for the eventual replacement of the meteor shield station objective which I see complained about often
4. All of our shield generators are either outdated or are meant to contain something, the only shield in the game that behaves anywhere near like this is the one that appears when drawing the nar`sie summon rune as a cultist
## TO-DO

1. Finish coding the node (done)
2. Sprite the charger/node/relay/well (done)
3. Make sure they play well with power/lack of power (seems fine so far)
4. Make the boards obtainable (done)
5. Finalize the tgui of the main generator (done)
6. Do a lot of cleaning up and optimizing, especially with slowing down the generation of the forcefield since it doesn`t really need to be instant (done???)
7. Test for nonobvious cases that might break the game (tram harmlessly deletes the forcefield so it`s a safe interaction) (it looks safe to me to far, it interacts with immovable rod and meteors as you should expect)


## Changelog
:cl:
add: Nanotrasen has uncovered a unique use for plasteel allowing the construction of new machines capable of fabricating and distributing hardlight into a modular barrier. These new machines are available via the advanced power manipulation techweb.

/:cl:

moocowswag committed:
Update shieldgen.dm

moocowswag committed:
Update shieldgen.dm

moocowswag committed:
can we build it?

moocowswag committed:
firsttest

moocowswag committed:
works a bit

moocowswag committed:
shieldtakesdamage

moocowswag committed:
alphaupdates

moocowswag committed:
startingonui

moocowswag committed:
ifinallymadedafile

moocowswag committed:
Bar???

moocowswag committed:
ihatetguiihatetguiihatetguiihatetgui

moocowswag committed:
Update shieldgen.dm

moocowswag committed:
thatwasfunnowfixesandnerfs

moocowswag committed:
cleanup nerfs faster code

moocowswag committed:
nowtakesdamagefromatmos

moocowswag committed:
cleanups and starting on modular stuff

moocowswag committed:
wires

moocowswag committed:
huh?

moocowswag committed:
some more cleaning and they connect!

moocowswag committed:
testin

moocowswag committed:
justsomemorecleaningsoitseasiertotest

moocowswag committed:
testing sprite proportions

moocowswag committed:
first sprites?

moocowswag committed:
mainspritesdone?

moocowswag committed:
reee

moocowswag committed:
grrrr

moocowswag committed:
overload is a more appropriate name

moocowswag committed:
more tgui work and toggling exterior shields

moocowswag committed:
interior-cost-and-fixes

moocowswag committed:
whyamibalancingitbeforefinishingit

moocowswag committed:
workonconnectingmachines

moocowswag committed:
we`ve reached modularity

moocowswag committed:
increasedpartcount

moocowswag committed:
fixedabug

moocowswag committed:
Merge branch 'master' into ShieldGeneratorModular

moocowswag committed:
why were they renamed to servos i havent slept in so long

moocowswag commented:
Ill work on this pr more tomorrow, another thing to add is its meant to be sort of a side grade to reinforced walls with upsides and downsides

moocowswag committed:
almostthere

moocowswag committed:
incaseadmindoesfunny

moocowswag commented:
github is taking a poopie my commits are being sent to the shadow realm and i cant update this pr to the latest master

moocowswag committed:
bugfix,nodeschainconnect

moocowswag committed:
thechaindisconnects

moocowswag committed:
qol, range scaling nerf, fix,

moocowswag committed:
Merge branch 'master' into ShieldGeneratorModular

moocowswag commented:
Poopie time ended, node is mostly finished it just needs qol changes.

moocowswag committed:
doesthisfixtest

moocowswag committed:
NOW its fixed

moocowswag committed:
fixmoretests

moocowswag committed:
fixremainingtests?

moocowswag committed:
Merge branch 'master' into ShieldGeneratorModular

moocowswag commented:
all checks passed, continuing to generate shitcode

moocowswag committed:
nodes done!!!

moocowswag committed:
sprites

moocowswag committed:
uhh

moocowswag committed:
nodespritedone

moocowswag committed:
bugfixes

moocowswag committed:
opps

moocowswag commented:
its in a playable state now just gotta sprite it then test power and other stuff

moocowswag committed:
wellsprites

moocowswag committed:
fuck

moocowswag committed:
ugh

moocowswag committed:
chargersprites?

moocowswag committed:
relaysprites

moocowswag committed:
happywithradiusscalingnowalsospriteworks

moocowswag commented:
all of the sprites are complete, the radius scaling from parts was nerfed to 0.25 and im happy with the equation that rewards you for using a small shield and makes you pay for a large one.

moocowswag committed:
forcefieldnowgoesdownwhennopower

moocowswag committed:
added to adv power manip techweb and is printable

moocowswag commented:
fully available and playable, sprites are done, the boards and machines are achievable without debugging

moocowswag committed:
ineverwannatouchauieveragain

moocowswag committed:
empisrareandhardtodoconsistentlysoitshouldbestronger

moocowswag committed:
Merge branch 'master' into ShieldGeneratorModular

moocowswag commented:
Time for the review/optimize/cleanup period

moocowswag committed:
noregenwhennopower

moocowswag committed:
couldntfindtaskbuildallsoihadtocommitthelastchangebeforetestingit

moocowswag committed:
heavyoptimizationtolargeshields

moocowswag commented:
performance shouldnt be a big concern with large shields now, especially with how hard they are to accomplish.

moocowswag committed:
Merge branch 'master' into ShieldGeneratorModular

moocowswag committed:
thatsnot10

moocowswag committed:
Merge branch 'ShieldGeneratorModular' of https://github.com/moocowswag/FeatureInflation13 into ShieldGeneratorModular

moocowswag committed:
imtootired

san7890 commented:
i just added the draft state since you still have (WIP) in your PR title, feel free to mark "Ready for review" and remove the `(WIP)` from your title when this is in a mergeable state

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
 #undef ACTIVE_HASFIELDS
+
+//Modular Shield Generator Start
+/obj/machinery/modular_shield_gen
```
```suggestion
/obj/machinery/modular_shield_generator
```

there's no reason to abbreviate.

moocowswag replied:
you underestimate the amount of times i`ll have to rewrite that guess ill get started

moocowswag replied:
done

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+	var/list/inside_shield = null
+
+	///The list of machines that are connected to and boosting us
+	var/list/obj/machinery/modular_shield/module/connected_modules = null
```
are you doing this like a lazylist? If you don't know what that is, i would instead do this:

```suggestion
	var/list/obj/machinery/modular_shield/module/connected_modules = list()
```

this is just so we can keep it typed as a list for list-specific procs and variables like `len` and whatever

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+	///Max strength gained from our own parts
+	var/innate_strength
+
+	var/list/list_of_turfs
```
dmdoc this, and also instantiate the list like `var/list/list_of_turfs = list()`

moocowswag replied:
if dmdoc means /// then done

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+/obj/machinery/modular_shield_gen/RefreshParts()
+	. = ..()
+
+	innate_regen = 3
+	innate_radius = 3
+	innate_strength = 40
```
These should be on the subtype var-declarations itself, and then if you want to "reset" the variables somehow, you can do something like:

```suggestion
	innate_regen = initial(innate_regen)
	innate_radius = initial(innate_radius)
	innate_strength = initial(innate_strength)
```

`initial()` is pretty neat because it will automatically set the variable back to what it was at compile time.

moocowswag replied:
recommited idk why it didnt save the first time

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+
+/datum/wires/modular_shield_gen/New(atom/holder)
+	wires = list(WIRE_HACK)
+	..()
```
```suggestion
	return ..()
```

this syntax is preferred.

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+		if(WIRE_HACK)
+			G.toggle_shields()
+			return
+	..()
```
```suggestion
	return ..()
```

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+
+/datum/wires/modular_shield_gen/on_pulse(wire)
+
+	var/obj/machinery/modular_shield_gen/G = holder
```
no single-letter variables in new code unless it's really super-duper scorching hot, and this will never be that level of hot.

moocowswag replied:
renamed to shield_gen resolving

moocowswag commented:
> i just added the draft state since you still have (WIP) in your PR title, feel free to mark "Ready for review" and remove the `(WIP)` from your title when this is in a mergeable state

 its playable ive been testing it in game and played like i played a normal round and used it, the reason why i marked as ready for review was because most of the code is final and i want people to actually look at it and tell me if its not up to standard or something.

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+			return
+	..()
+
+/obj/machinery/modular_shield_gen/proc/deactivate_shields()
```
you should be DMdoccing all the new procs you're adding, if for no other reason but to help statistics. https://github.com/tgstation/tgstation/blob/master/.github/guides/AUTODOC.md

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+			deployed_shields += deploying_shield
+	calculate_regeneration()
+	active_power_usage += deployed_shields.len * BASE_MACHINE_ACTIVE_CONSUMPTION * 0.1
+	return
```
```suggestion
```

redundant.

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+		var/dx = fradius
+		var/dy = 0
+		var/t2
+		var/list/list_of_turfs = list()
+		while(dx >= dy)
+			list_of_turfs += locate(x + dx, y + dy, z)
+			list_of_turfs += locate(x - dx, y + dy, z)
+			list_of_turfs += locate(x + dx, y - dy, z)
+			list_of_turfs += locate(x - dx, y - dy, z)
+			list_of_turfs += locate(x + dy, y + dx, z)
+			list_of_turfs += locate(x - dy, y + dx, z)
+			list_of_turfs += locate(x + dy, y - dx, z)
+			list_of_turfs += locate(x - dy, y - dx, z)
+			dy += 1
+			t1 += dy
+			t2 = t1 - dx
+			if(t2 > 0)
+				t1 = t2
+				dx -= 1
```
what the fuck? look for stuff in places like `/code/_HELPERS/maths.dm` for procs that do stuff like this for you already, there has to be something like `get_angle` or whatever you need for this. if there isn't something already, i would rather you add this part of the code to `maths.dm` so it can at least be used to help other people hsould they really truly need it

moocowswag replied:
there is a reason why i made this only run when the shield radius gets very large, it generates a pretty ugly circle when its smaller and it takes decimals very horribly badly.

ill look into that though.

san7890 replied:
you can have it read the radius and then alternate between two procs

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+			toggle_shields()
+			return
+		if ("toggle_exterior")
+			exterior_only = !(exterior_only)
```
```suggestion
			exterior_only = !exterior_only
```

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+			radius = max(1,(text2num(params["new_radius"])))
+		if ("toggle_shields")
+			toggle_shields()
+			return
```
```suggestion

```

not useful because there's no more possible logic that we should avoid executing

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+			return
+		if ("toggle_exterior")
+			exterior_only = !(exterior_only)
+			return
```
```suggestion

```

ditto redudnant/not useful

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+//calculations for the shield`s core stats
+/obj/machinery/modular_shield_gen/proc/calculate_boost()
+
+	regen_boost = 0
+	for (var/obj/machinery/modular_shield/module/charger/new_charger in connected_modules)
+		regen_boost += new_charger.charge_boost
+
+	calculate_regeneration()
+
+	max_strength_boost = 0
+	for (var/obj/machinery/modular_shield/module/well/new_well in connected_modules)
+		max_strength_boost += new_well.strength_boost
+
+	calculate_max_strength()
+
+	radius_boost = 0
```
ditto on these three vars going back to being tied to the type declaration itself and using `initial()` memes as you need it.

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+
+	max_regeneration = innate_regen + regen_boost
+
+	if(!(active))
```
```suggestion
	if(!active)
```

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+			return
+		current_regeneration = max_regeneration
+		return
+//we lose more than half the regeneration rate when generating a shield that is near the max radius that we can handle but if we generate
+//a shield with a very small fraction of the max radius we can support we get a very small bonus multipliar
+	current_regeneration = (max_regeneration / (0.5 + (radius * 2)/max_radius))
```
```suggestion
	//we lose more than half the regeneration rate when generating a shield that is near the max radius that we can handle but if we generate
	//a shield with a very small fraction of the max radius we can support we get a very small bonus multiplier
	current_regeneration = (max_regeneration / (0.5 + (radius * 2)/max_radius))
```

looks better when it's all inline. also small spelling fix

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+	current_regeneration = (max_regeneration / (0.5 + (radius * 2)/max_radius))
+
+	if(!exterior_only)
+		current_regeneration *=0.5
```
```suggestion
		current_regeneration *= 0.5
```

spacing nit

moocowswag replied:
done

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+
+
+//Start of other machines
+/obj/machinery/modular_shield/module //The general code used for machines that want to connect to the network
```
dmdoc this

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+
+	connected_turf = get_step(loc, dir)
+
+/obj/machinery/modular_shield/module/Destroy()
+	. = ..()
+	if(shield_generator)
+		shield_generator.connected_modules -= (src)
+		shield_generator.calculate_boost()
+	if(connected_node)
+		connected_node.connected_through_us -= (src)
+	return ..()
```
do not invoke parent twice here. you probably meant

```suggestion
/obj/machinery/modular_shield/module/Destroy()
	if(shield_generator)
		shield_generator.connected_modules -= (src)
		shield_generator.calculate_boost()
	if(connected_node)
		connected_node.connected_through_us -= (src)
	return ..()
```

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+//rather than automatically checking for connections its probably alot less expensive to just make the players manually multi tool sync each part
+	if(I.tool_behaviour == TOOL_MULTITOOL)
+		try_connect(user)
+		return
+
+	if(default_change_direction_wrench(user, I))
+		if(shield_generator)
+			shield_generator.connected_modules -= (src)
+			shield_generator.calculate_boost()
+			shield_generator = null
+			update_icon_state()
+		if(connected_node)
+			connected_node.connected_through_us -= (src)
+			connected_node = null
+		connected_turf = get_step(loc, dir)
+		return
+
+	if(default_deconstruction_crowbar(I))
+		return
```
these should have return types, like `return TRUE` or `return FALSE`

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+/obj/machinery/modular_shield/module/proc/try_connect(user)
+
+	if(shield_generator || connected_node)
+		balloon_alert(user, "already connected to something")
```
```suggestion
		balloon_alert(user, "already connected to something!")
```

error states should have exclamation points.

san7890 replied:
bump

moocowswag replied:
commited manually

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+	if(shield_generator)
+
+		shield_generator.connected_modules += (src)
+		balloon_alert(user, "connected directly to generator")
```
```suggestion
		balloon_alert(user, "connected to generator")
```

directly is superfluous here imo

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+		shield_generator = connected_node.shield_generator
+		if(shield_generator)
+			shield_generator.connected_modules += (src)
+			balloon_alert(user, "connected to generator through node")
+			update_icon_state()
+			if(istype(src, /obj/machinery/modular_shield/module/node))
+				var/obj/machinery/modular_shield/module/node/connected_node = src
+				connected_node.connect_connected_through_us()
+			shield_generator.calculate_boost()
+			return
+		balloon_alert(user, "connected to node but no path to generator")
+		return
+	balloon_alert(user, "failed to detect powered generator or node")
```
these balloon alerts are way too long, make them more succinct or change it to a `to_chat`.

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+/obj/machinery/modular_shield/module/node
+
+	name = "Modular Shield Node"
+	desc = "A waist high mess of humming pipes and wires that extend the modular shield network"
```
```suggestion
	desc = "A waist high mess of humming pipes and wires that extend the modular shield network."
```

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+	active_power_usage = BASE_MACHINE_ACTIVE_CONSUMPTION * 0.5
+	circuit = /obj/item/circuitboard/machine/modular_shield_node
+
+	var/list/connected_through_us
```
dmdoc

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+		return
+	icon_state = "node_on_[panel_open ?"open":"closed"]"
+
+/obj/machinery/modular_shield/module/node/Initialize(mapload)
+	. = ..()
+	connected_through_us = list()
```
you can just set the whole `list()` thing on the type declaration and skip having this `Initialize()` override.

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+
+/obj/machinery/modular_shield/module/node/update_icon_state()
+	. = ..()
+	if((!shield_generator) || (machine_stat & NOPOWER))
```
```suggestion
	if(!shield_generator || (machine_stat & NOPOWER))
```

san7890 commented on tgui/packages/tgui/interfaces/Modularshieldgen.tsx:
```diff
@@ -0,0 +1,147 @@
+import { Window } from '../layouts';
+import { Stack, Section, ProgressBar, Button, NumberInput } from '../components';
+
+type Modularshieldgendata = {
```
I believe JS wants you to do camelCase for everything (snake_case is fine for stuff imported from DM) so this should be `modularShieldGeneratorData` instead.

san7890 commented on tgui/packages/tgui/interfaces/Modularshieldgen.tsx:
```diff
@@ -0,0 +1,147 @@
+  exterior_only: Boolean;
+};
+
+export const Modularshieldgen = (props, context) => {
```
keeping in line with JS stuff, `modularShieldGenerator`.

moocowswag replied:
it says my ui is missing an export when i change that

san7890 replied:
sorry i meant `ModularShieldGen` in this instance (edit: there's another part of the review that i think i found the issue with?)

moocowswag replied:
i tried changing both, when i changed both to camel case it said there was no "modularShieldGen" thingy so im assuming it has to do with the name of the file

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,644 @@
+
+/obj/machinery/modular_shield/module/well/update_icon_state()
+	. = ..()
+	if((!shield_generator) || (machine_stat & NOPOWER))
```
```suggestion
	if(!shield_generator || (machine_stat & NOPOWER))
```

san7890 reviewed (changes requested):
there's probably way more stuff that should be patched out over time but have 30 nits to deal with whenever you get the chance. i would strongly advise to do each type of review (i.e. do all the dmdocs at once, do all the `initial()` memes at once) as one commit per-type so that if i'm going senile it's very easy to revert a bad suggestion and retain fully working code. beauty of git

if i completely misread something feel free to just write a few sentences (no real need for paragraphs unless stuff is really confusing) on why it needs to be that way and we can go from there.

ask in #coding-general if you need help, i'm not really available through DMs like I once was. there's a good link on how to ask questions in the topic of that channel.

moocowswag commented:
I`ll get to that after i test these edge cases, it plays well with immovable rod! (the rod goes through the shield and instantly empties the generator`s health pool which seems fair)

moocowswag commented:
It blocks and properly takes damage from meteors!

moocowswag committed:
erator erator erator erator erator erator erator erator

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: san7890 <the@san7890.com>

moocowswag commented:
commited and pulled most of the safe looking code, will do alot more tomorrow.

moocowswag committed:
lists, initial vals, balloons, dmdocs

moocowswag committed:
Merge branch 'ShieldGeneratorModular' of https://github.com/moocowswag/FeatureInflation13 into ShieldGeneratorModular

moocowswag commented:
uhhh im just gonna manually do the suggested commits on my own machine because when i tried to pull from github it gave me like a whole page of red errors including stuff i didnt touch.

san7890 commented:
> uhhh im just gonna manually do the suggested commits on my own machine because when i tried to pull from github it gave me like a whole page of red errors including stuff i didnt touch and has been in this file long before this pr.

that's okay

moocowswag committed:
thecamelthatworks

moocowswag committed:
Update shieldgen.dm

moocowswag commented:
i think i burned through the majority of requested changes... ill look into math.dm but i doubt ill find anything there that is better than this and i don`t think the midpoint algorithm should be added to it because it has a lot of "catches" or things you have to worry about when using it.

moocowswag committed:
testedandadminproofed

moocowswag commented:
okay i looked through math dm and while i did find something there that i will use in the future (specifically the get_line proc) i dont see anything else there that resembles getting the perimeter of a circle.

 the midpoint algorithm seems to be the most efficient way we have of doing it when grabbing the perimeter of turfs of a large circle, my main issue is how i should implement it into math dm or if i should even do that because it doesn't seem to play well in certain conditions which i had to make sure to avoid with my own code.

moocowswag committed:
idontknowwhybutthesechangeswentmissing

moocowswag commented:
changes from before went missing and had to readd them, went ahead and did more testing, it plays well with power again i just spent 30 minutes testing and building the machine, its still accessible via research then printing, it stops regenerating and the shield is erased when power is lost, etc.

All thats left really is what to do with that midpoint formula and possible balance changes.

moocowswag committed:
Merge branch 'master' into ShieldGeneratorModular

moocowswag committed:
thesegotrevertedtootestedandrecommited

moocowswag committed:
Merge branch 'ShieldGeneratorModular' of https://github.com/moocowswag/FeatureInflation13 into ShieldGeneratorModular

moocowswag committed:
spellingstuff

san7890 commented:
> All thats left really is what to do with that midpoint formula and possible balance changes and ofcourse more reviews.

if you can genericize it to spit out raw values and then have the code accept those raw values and do whatever edge-casey stuff that could potentially work. i would start out by transplanting that whole segment of the code into its own proc, and then seeing how much you can strip out into an agnostic proc, and then seeing how you can work for the data.

the equation just needs to be an equation, right? so just have it so that every run invokes the checkable equation, gets the raw values (probably in a list or a `json_encode` or something) and then we can decode that list and work with the data that we get out.

san7890 commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +159,50 @@
+	name = "Modular Shield Generator"
+	greyscale_colors = CIRCUIT_COLOR_ENGINEERING
+	build_path = /obj/machinery/modular_shield_generator
+	req_components = list(
+		/datum/stock_part/servo =	1,
+		/datum/stock_part/micro_laser = 1,
+		/datum/stock_part/capacitor = 1,
+		/obj/item/stack/sheet/plasteel = 3)
```
this is a wack ass way to write a list

```suggestion
	req_components = list(
		/datum/stock_part/servo = 1,
		/datum/stock_part/micro_laser = 1,
		/datum/stock_part/capacitor = 1,
		/obj/item/stack/sheet/plasteel = 3,
	)
```

san7890 commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +159,50 @@
+	name = "Modular Shield Node"
+	greyscale_colors = CIRCUIT_COLOR_ENGINEERING
+	build_path = /obj/machinery/modular_shield/module/node
+	req_components = list(
+	/obj/item/stack/cable_coil = 15,
+	/obj/item/stack/sheet/plasteel = 2)
```
```suggestion
	req_components = list(
	/obj/item/stack/cable_coil = 15,
	/obj/item/stack/sheet/plasteel = 2,
	)
```

san7890 commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +159,50 @@
+	name = "Modular Shield Well"
+	greyscale_colors = CIRCUIT_COLOR_ENGINEERING
+	build_path = /obj/machinery/modular_shield/module/well
+	req_components = list(
+			/datum/stock_part/capacitor = 3,
+			/obj/item/stack/sheet/plasteel = 2)
```
```suggestion
	req_components = list(
		/datum/stock_part/capacitor = 3,
		/obj/item/stack/sheet/plasteel = 2,
	)
```

san7890 commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +159,50 @@
+	name = "Modular Shield Relay"
+	greyscale_colors = CIRCUIT_COLOR_ENGINEERING
+	build_path = /obj/machinery/modular_shield/module/relay
+	req_components = list(
+		/datum/stock_part/micro_laser = 3,
+		/obj/item/stack/sheet/plasteel = 2)
```
```suggestion
	req_components = list(
		/datum/stock_part/micro_laser = 3,
		/obj/item/stack/sheet/plasteel = 2,
	)
```

san7890 commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +159,50 @@
+	name = "Modular Shield Charger"
+	greyscale_colors = CIRCUIT_COLOR_ENGINEERING
+	build_path = /obj/machinery/modular_shield/module/charger
+	req_components = list(
+		/datum/stock_part/servo = 3,
+		/obj/item/stack/sheet/plasteel = 2)
```
```suggestion
	req_components = list(
		/datum/stock_part/servo = 3,
		/obj/item/stack/sheet/plasteel = 2,
	)
```

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+	///This is the list of perimeter turfs that we grab when making large shields of 10 or more radius
+	var/list/list_of_turfs = list()
+
+	processing_flags = START_PROCESSING_ON_INIT
```
old vars go before new vars, move this before all the new `var/x` declarations.

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+
+/obj/machinery/modular_shield_generator/power_change()
+	. = ..()
+	if(machine_stat & NOPOWER)
+		deactivate_shields()
+		end_processing()
+		return
+	begin_processing()
```
```suggestion
	if(!(machine_stat & NOPOWER))
		begin_processing()
		return
		
	deactivate_shields()
	end_processing()
```

cleaner organization to my eyes

moocowswag replied:
bruh

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+			return
+
+		for(var/turf/target_tile as anything in list_of_turfs)
+			if (!(target_tile in inside_shield) && isopenturf(target_tile) && !(locate(/obj/structure/emergency_shield/modular) in target_tile))
```
```suggestion
			if (isopenturf(target_tile) && !(target_tile in inside_shield) && !(locate(/obj/structure/emergency_shield/modular) in target_tile))
```

this is better because the typecheck is less expensive than the locate/in stuff, and the placement means that we don't have to waste time doing the locate/in stuff if the turf wasn't even valid to begin with. reads nicer too.

san7890 replied:
bump

moocowswag replied:
done

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+	var/list/inside_shield = circle_range_turfs(src, radius - 1)//in the future we might want to apply an effect to the turfs inside the shield
+	if(exterior_only)
+		for(var/turf/target_tile as anything in circle_range_turfs(src, radius))
+			if (!(target_tile in inside_shield) && isspaceturf(target_tile) && !(locate(/obj/structure/emergency_shield/modular) in target_tile))
```
```suggestion
			if (isspaceturf(target_tile) && !(target_tile in inside_shield) && !(locate(/obj/structure/emergency_shield/modular) in target_tile))
```

san7890 replied:
bump

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+		return
+
+	for(var/turf/target_tile as anything in circle_range_turfs(src, radius))
+		if (!(target_tile in inside_shield) && isopenturf(target_tile) && !(locate(/obj/structure/emergency_shield/modular) in target_tile))
```
```suggestion
		if (isopenturf(target_tile) && !(target_tile in inside_shield) && !(locate(/obj/structure/emergency_shield/modular) in target_tile))
```

san7890 replied:
bump

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+
+/obj/machinery/modular_shield_generator/update_icon_state()
+
+	icon_state = ("gen_[!(machine_stat & NOPOWER)?"[recovering ?"recovering_":"ready_"]":"no_power_"][(panel_open)?"open":"closed"]")
```
this is really wack and it's because you don't have spaces outside the quotes, just do it like this:

```suggestion
	icon_state = ("gen_[!(machine_stat & NOPOWER) ? "[recovering ? "recovering_" : "ready_"]" : "no_power_"][(panel_open) ? "open" : "closed"]")
```

this won't (or shouldn't) introduce weird spaces into the icon_state if i didn't screw anything up

moocowswag replied:
they still work

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+	. = ..()
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Modularshieldgen")
```
yeah this is why that UI thing was failing, it's because you didn't update it here?

moocowswag replied:
when i changed that to camel case it said it couldnt find a ui named modularShieldGen

when i only changed the export to camel case it said Modularshieldgen had no export.

moocowswag replied:
found a proper solution, leaving it unresolved incase you dont like it though.

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+	icon = 'icons/mecha/mech_bay.dmi'
+	icon_state = "recharge_port"
+	density = TRUE
+	dir = SOUTH
```
isn't it south by default?

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+	density = TRUE
+	dir = SOUTH
+
+	var/obj/machinery/modular_shield_generator/shield_generator
+
+	var/obj/machinery/modular_shield/module/node/connected_node
+
+	var/turf/connected_turf
```
dmdoc these

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+		to_chat(user, span_notice("You close the maintenance hatch of [src]."))
+		return TRUE
+
+//rather than automatically checking for connections its probably alot less expensive to just make the players manually multi tool sync each part
+	if(I.tool_behaviour == TOOL_MULTITOOL)
```
```suggestion
	//rather than automatically checking for connections its probably alot less expensive to just make the players manually multi tool sync each part
	if(I.tool_behaviour == TOOL_MULTITOOL)
```

did this regress?

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+			return
+		balloon_alert(user, "connected to node")
+		return
+	balloon_alert(user, "failed to find connection")
```
```suggestion
	balloon_alert(user, "failed to find connection!")
```

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+/obj/machinery/modular_shield/module/node/update_icon_state()
+	. = ..()
+	if(!shield_generator || (machine_stat & NOPOWER))
+		icon_state = "node_off_[panel_open ?"open":"closed"]"
+		return
+	icon_state = "node_on_[panel_open ?"open":"closed"]"
```
```suggestion
		icon_state = "node_off_[panel_open ?"open" : "closed"]"
		return
	icon_state = "node_on_[panel_open ?"open" : "closed"]"
```

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+	. = ..()
+
+	disconnect_connected_through_us()
+	if(shield_generator)
+		shield_generator.connected_modules -= (src)
+		shield_generator.calculate_boost()
+		shield_generator = null
+		update_icon_state()
```
```suggestion
	if(isnull(shield_generator))
		return

	shield_generator.connected_modules -= (src)
	shield_generator.calculate_boost()
	shield_generator = null
	update_icon_state()
```

doing this makes the intent of this proc much more explicit while leveraging the speed and powah of `isnull()`

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+
+/obj/machinery/modular_shield/module/charger/update_icon_state()
+	. = ..()
+	if(!shield_generator || (machine_stat & NOPOWER))
```
```suggestion
	if(isnull(shield_generator) || (machine_stat & NOPOWER))
```

this is like 0.0001 cpu cycles faster, and it's easier to read. win-win.

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+
+/obj/machinery/modular_shield/module/relay/update_icon_state()
+	. = ..()
+	if(!shield_generator || (machine_stat & NOPOWER))
```
ditto isnull

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+
+/obj/machinery/modular_shield/module/well/update_icon_state()
+	. = ..()
+	if(!shield_generator || (machine_stat & NOPOWER))
```
ditto isnull

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+	return exposed_temperature > (T0C + 400) //starts taking damage from high temps at the same temperature that nonreinforced glass does
+
+/obj/structure/emergency_shield/modular/atmos_expose(datum/gas_mixture/air, exposed_temperature)
+	if(shield_generator)
+		shield_generator.shield_drain(round(air.return_volume() / 400))//400 integer determines how much damage the shield takes from hot atmos (higher value = less damage)
+		return
+	qdel(src)
```
```suggestion
	if(isnull(shield_generator))
		qdel(src)
		return
	
	shield_generator.shield_drain(round(air.return_volume() / 400))//400 integer determines how much damage the shield takes from hot atmos (higher value = less damage)
```

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,656 @@
+/obj/structure/emergency_shield/modular/take_damage(damage_amount, damage_type = BRUTE, damage_flag = 0, sound_effect = 1, attack_dir)
+	. = ..()
+	if(damage_type == BRUTE || damage_type == BURN)
+		if(shield_generator)
+			shield_generator.shield_drain(damage_amount)//can add or subtract a flat value to buff or nerf crowd damage
+			return
+		qdel(src)
+
+/obj/structure/emergency_shield/modular/emp_act(severity)
+	if(shield_generator)
+		shield_generator.shield_drain(15 / severity) //Light is 2 heavy is 1, note emp is usually a large aoe, tweak the number if not enough damage
+		return
+	qdel(src)
```
invert these so that we do the same `isnull(shield_generator)` check and qdel in that tree like i showed in the line 1189 review of this file.

san7890 reviewed (changes requested):
some things that i'm almost certain i marked were untouched and resolved, might be a consequence of some stuff deleting itself like you said? I just remarked it/re-opened it as i saw stuff. i wouldn't rely on github webedits as much this time, by-hand would be optimal.

sorry to dump another 25+ review on you again but it's important stuff

moocowswag commented:
> > All thats left really is what to do with that midpoint formula and possible balance changes and ofcourse more reviews.
> 
> if you can genericize it to spit out raw values and then have the code accept those raw values and do whatever edge-casey stuff that could potentially work. i would start out by transplanting that whole segment of the code into its own proc, and then seeing how much you can strip out into an agnostic proc, and then seeing how you can work for the data.
> 
> the equation just needs to be an equation, right? so just have it so that every run invokes the checkable equation, gets the raw values (probably in a list or a `json_encode` or something) and then we can decode that list and work with the data that we get out.

uhhhh i was just gonna put it next to the get_line proc which gives a list of turfs based on 2 turfs and name it get_perimeter which will accept a radius, an origina, and z level and return a perimeter of turfs.

moocowswag committed:
partsandpower

moocowswag commented:
commited and resolving parts & power change proc stuff, hopefully it saves.

moocowswag committed:
power and old flag or var thing

moocowswag committed:
autodocs

moocowswag committed:
balloonandcomment

moocowswag committed:
if(isnull(shield_generator)

moocowswag committed:
isnullisnullisnullisnullisnull

moocowswag committed:
isnullisnullisnullisnullisnullisnull

moocowswag commented:
Okay so i went into other files , and from the list of other interfaces in the same folder that Modularshieldgen is in youll see stuff like "AirlockController" with same export "AirlockController  "Biogenerator" with same export "Biogenerator" regardless if they do camel case or not the export needs to match the name of the file and its case sensitive

moocowswag committed:
test

moocowswag committed:
Rename Modularshieldgen.tsx to ModularShieldGen.tsx

moocowswag commented:
for some reason github desktop didnt recognize the rename of the file so hopefully itl let me pull and test it from github after i did a web edit

moocowswag commented:
pulled and tested, it works.

moocowswag committed:
iconstatesstillwork

moocowswag committed:
ifwiresdontworkblamethis

moocowswag committed:
movedalgorithmtomath_dm

moocowswag commented:
the algorithm works for us while chained to the confines of math helpers

san7890 commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -149,7 +149,8 @@
 		/datum/stock_part/servo = 2,
 		/obj/item/stack/cable_coil = 1,
-		/datum/stock_part/filter = 1)
+		/datum/stock_part/filter = 1
+		)
```
```suggestion
		/datum/stock_part/filter = 1,
	)
```

there's a very specific way to multiline lists, and that's that you need both a trailing comma and the indentation level of the closing parenthesis must match the starting indentation of the list variable

moocowswag replied:
what the, is a filter none of my machines use that

moocowswag replied:
oh lol tcomms parts okay

san7890 commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +160,53 @@
+		/datum/stock_part/servo = 1,
+		/datum/stock_part/micro_laser = 1,
+		/datum/stock_part/capacitor = 1,
+		/obj/item/stack/sheet/plasteel = 3
+		)
```
```suggestion
		/obj/item/stack/sheet/plasteel = 3,
	)
```

moocowswag replied:
> some stuff slipped you by, i just replied to any un-addressed suggestions with "bump"

thanks its getting hard keeping track of these ontop of the ones i left open to see if you were happy with them or not.

san7890 commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +160,53 @@
+	build_path = /obj/machinery/modular_shield/module/node
+	req_components = list(
+	/obj/item/stack/cable_coil = 15,
+	/obj/item/stack/sheet/plasteel = 2
```
```suggestion
	/obj/item/stack/sheet/plasteel = 2,
```

san7890 commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +160,53 @@
+	name = "Modular Shield Well"
+	greyscale_colors = CIRCUIT_COLOR_ENGINEERING
+	build_path = /obj/machinery/modular_shield/module/well
+	req_components = list(
+			/datum/stock_part/capacitor = 3,
+			/obj/item/stack/sheet/plasteel = 2
+			)
```
```suggestion
	req_components = list(
		/datum/stock_part/capacitor = 3,
		/obj/item/stack/sheet/plasteel = 2,
	)
```

san7890 commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +160,53 @@
+	name = "Modular Shield Relay"
+	greyscale_colors = CIRCUIT_COLOR_ENGINEERING
+	build_path = /obj/machinery/modular_shield/module/relay
+	req_components = list(
+		/datum/stock_part/micro_laser = 3,
+		/obj/item/stack/sheet/plasteel = 2
+		)
```
```suggestion
	req_components = list(
		/datum/stock_part/micro_laser = 3,
		/obj/item/stack/sheet/plasteel = 2,
	)
```

san7890 commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +160,53 @@
+	name = "Modular Shield Charger"
+	greyscale_colors = CIRCUIT_COLOR_ENGINEERING
+	build_path = /obj/machinery/modular_shield/module/charger
+	req_components = list(
+		/datum/stock_part/servo = 3,
+		/obj/item/stack/sheet/plasteel = 2
+		)
```
```suggestion
	req_components = list(
		/datum/stock_part/servo = 3,
		/obj/item/stack/sheet/plasteel = 2,
	)
```

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,643 @@
+		if(WIRE_HACK)
+			shield_gen.toggle_shields()
+
+	return..()
```
```suggestion
	return ..()
```

spooky otherwise

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,643 @@
+
+/datum/wires/modular_shield_generator/New(atom/holder)
+	wires = list(WIRE_HACK)
+	return..()
```
```suggestion
	return ..()
```

san7890 commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,643 @@
+
+	if(default_deconstruction_crowbar(I))
+		return TRUE
+	return..()
```
```suggestion
	return ..()
```

san7890 reviewed (changes requested):
some stuff slipped you by, i just replied to any un-addressed suggestions with "bump"

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,147 @@
+  current_regeneration: number;
+  max_radius: number;
+  current_radius: number;
+  active: Boolean;
+  recovering: Boolean;
+  exterior_only: Boolean;
```
BooleanLike

moocowswag replied:
"cannot find name BooleanLike" errors

jlsnow301 replied:
need to import BooleanLike. you should be able to hover over it / get a blue light bulb icon

jlsnow301 replied:
![Screenshot 2023-05-17 152710](https://github.com/tgstation/tgstation/assets/42397676/64fb2520-1439-43c4-a5ae-76c67d10c17b)

moocowswag replied:
thanks, i dont get lightbulbs, i looked at the autolathe interface code and saw how it imports BooleanLike and used that.

jlsnow301 replied:
which editor are you using?

moocowswag replied:
visual studio code

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,147 @@
+import { Window } from '../layouts';
+import { Stack, Section, ProgressBar, Button, NumberInput } from '../components';
+
+type modularShieldGeneratorData = {
```
in local files, I think short names are fine. Data is totally acceptable.
the other issue with this is that (by convention, no real rule,) types are PascalCase.

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,147 @@
+                  Radius {current_radius}/{max_radius}
+                </ProgressBar>
+              </Section>
+              <Section title={'Settings'} bold={1}>
```
no need to make 'Settings' an object like this. some other instances where this is the case also. title="Settings", title="Set Radius", etc

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,147 @@
+                      }
+                    />
+
+                    <Section title={'Set Radius'} right={-10.5} top={-4.25}>
+                      Toggle Placement
+                    </Section>
+                    <Section right={-25} top={-7.35}>
+                      Toggle Shields
+                    </Section>
```
i'd really rather this information displayed inside of the button

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,147 @@
+                      right={4.25}
+                      top={3.25}
+                      onClick={() => act('toggle_exterior')}
+                      content={
+                        exterior_only ? 'External only' : 'Internal & External'
+                      }
```
`content` is deprecated as the way to write text inside of button, please just put the content as a child, `<Button ...>Text</Button>`

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,147 @@
+            <Section
+              horizontal
+              fill
+              height={'95px'}
+              title={'Regeneration and Radius'}>
+              <Section height={'20px'}>
+                <ProgressBar
+                  height={'20px'}
```
fill and height are conflicting terms. also, it may be unavoidable, but try using grow={xnumber} instead of defining the sizes in pixels. you'll save yourself some brain cells later.

jlsnow301 reviewed (changes requested):
i'm not thrilled with how the UI looks, it's functional but it subverts a lot of the standard theme and misuses some props. wrote comments below, it's serviceable

moocowswag committed:
emergencycommitincasethunderstormwrecksmymachineadddelay

moocowswag commented:
> i'm not thrilled with how the UI looks, it's functional but it subverts a lot of the standard theme and misuses some props. wrote comments below, it's serviceable

ya ill definitely take more looks at it again soon i was really really frustrated as it was my first time ever making a ui so alot of trial and error

jlsnow301 commented:
> ya ill definitely take more looks at it again soon i was really really frustrated as it was my first time ever making a ui so alot of trial and error so just got something that was semi presentable then moved on to other stuff

no problem! sadly with JS, there's so many ways to do things (and outdated guides) it can be difficult to find what works. reach out on discord if you need some help, happy to walk through it. good job using with typescript in your first ui btw

moocowswag committed:
tinybitofcleanup,afix,smallstuff

moocowswag commented:
Just did a small change to the behavior that I think will go a long way towards making this feature more healthy for the game,
now all of the shield tiles will have a lower density (and lower alpha to reflect that) for the first 2 seconds of their life time and up until these 2 seconds are up the generator cannot be manually turned off via ui or wire pulsing so spamming the shield on and off is no longer possible and the shield itself wont be able to surprise someone without warning.

moocowswag committed:
thiswasredundantbecauserelayalreadycostspower

removedactivepowerusage modifiers from the shield generator because their goal to make range cost power was already fulfilled by relays costing power, also why was it in the destroy proc lol....

moocowswag committed:
booleanlike

moocowswag committed:
brings tcomm boards up to standard

moocowswag committed:
no wait this one does that

moocowswag committed:
space between return and parentcall

moocowswag committed:
bruh i already did this

moocowswag committed:
exclamationpointonerrorballoon

moocowswag committed:
spacing

moocowswag committed:
whydidthatworkinthefirstplace,alsoisspaceturf(targettile)

moocowswag commented:
taking another full look at ui now, we are almost there!

moocowswag committed:
uirework

moocowswag commented:
![image](https://github.com/tgstation/tgstation/assets/62126254/13816292-9ec6-4046-ac97-bab538d820a7)

moocowswag committed:
reeee

moocowswag commented:
ignore the 999 max radius that was me manually changing the values so i could test triple digit radius on ui

moocowswag committed:
capitalizes the l

moocowswag committed:
oops

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+  return (
+    <Window title="Modular Shield Generator" width={600} height={400}>
+      <Window.Content scrollable>
+        <Section title={'Shield Strength'} color={recovering ? 'red' : 'white'}>
```
```suggestion
        <Section title="Shield Strength" color={recovering ? 'red' : 'white'}>
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+            value={current_strength}
+            minValue={0}
+            maxValue={max_strength}
+            height={'30px'}
```
```suggestion
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+            {current_strength}/{max_strength}
+          </ProgressBar>
+        </Section>
+        <Section title={'Regeneration and Radius'}>
```
```suggestion
        <Section title="Regeneration and Radius">
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+          <ProgressBar
+            title="Shield Strength"
+            value={current_strength}
+            minValue={0}
```
```suggestion
```
This should be default iirc

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+            height={'20px'}
+            title="Regeneration rate"
+            value={current_regeneration}
+            minValue={0}
```
```suggestion
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+        </Section>
+        <Section title={'Regeneration and Radius'}>
+          <ProgressBar
+            height={'20px'}
```
```suggestion
            height="20px"
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+          </ProgressBar>
+          <Section>
+            <ProgressBar
+              height={'20px'}
```
```suggestion
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+          </Section>
+        </Section>
+        <Section title={'Settings'}>
+          <Stack>
+            <Section>
```
```suggestion
```
These should not be trying to piece together sections like this

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+            <Section>
+              <LabeledList>
+                <LabeledList.Item label="Set Radius">
+                  <Section right={0.75}>
```
```suggestion
```

moocowswag replied:
how will i adjust the placement of the num input?

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+                    />
+                  </Section>
+                </LabeledList.Item>
+              </LabeledList>
+            </Section>
+            <Section>
+              <LabeledList>
```
```suggestion
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+                        })
+                      }
+                    />
+                  </Section>
```
```suggestion
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+            <Section>
+              <LabeledList>
+                <LabeledList.Item label="Placement Limitations">
+                  <Section right={0.75}>
```
```suggestion
```
Try to limit or remove any measurements like this. Let flex do the work for you

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+                      onClick={() => act('toggle_exterior')}>
+                      {exterior_only ? 'External only' : 'Internal & External'}
+                    </Button>
+                  </Section>
```
```suggestion
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+                    </Button>
+                  </Section>
+                </LabeledList.Item>
+              </LabeledList>
+            </Section>
+            <Section>
+              <LabeledList>
```
```suggestion
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+            <Section>
+              <LabeledList>
+                <LabeledList.Item label="Toggle Power">
+                  <Section right={0.75}>
```
```suggestion
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+                      icon="power-off"
+                      onClick={() => act('toggle_shields')}
+                    />
+                  </Section>
```
```suggestion
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+                  </Section>
+                </LabeledList.Item>
+              </LabeledList>
+            </Section>
+          </Stack>
```
```suggestion
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+
+  return (
+    <Window title="Modular Shield Generator" width={600} height={400}>
+      <Window.Content scrollable>
```
```suggestion
      <Window.Content>
          <Stack fill>
           <Stack.Item grow={2}>
```

experiment with different grow percentages to get the desired result

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+            </ProgressBar>
+          </Section>
+        </Section>
+        <Section title={'Settings'}>
```
```suggestion
         </ Stack.Item>
         <Stack.Item grow>
        <Section title="Settings">
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+            </Section>
+          </Stack>
+        </Section>
+      </Window.Content>
```
```suggestion
</ Stack.Item>
</ Stack>
      </Window.Content>
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,142 @@
+                      value={current_radius}
+                      minValue={3}
+                      maxValue={max_radius}
+                      lineHeight="30px"
+                      fontSize="15px"
+                      width="60px"
+                      height="20px"
```
i'd prefer these unedited but i will let it slide

moocowswag committed:
someuicleaning

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

jlsnow301 commented:
Sorry to bikeshed so hard over this, I'm picky with UIs, hoping we can get it looking nice

moocowswag commented:
![image](https://github.com/tgstation/tgstation/assets/62126254/fbd19968-09e3-4227-8fcb-19b9af5aa1a2)
you wanted this?

moocowswag committed:
horizontalui

jlsnow301 commented:
> keep in mind i was planning on there possibly being more settings in the future thats why i wanted it on the bottom but this work and can fit 2 or 3 additions which is probably plenty of future proofing.

if so, you can safely remove the stack and stack items.  toggle power should also be included in the settings section,

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,126 @@
+              <LabeledList>
+                <LabeledList.Item label="Set Radius">
+                  <NumberInput
+                    disabled={active}
```
```suggestion
                    disabled={active}
                    fluid
```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,126 @@
+                }}>
+                Regeneration {current_regeneration}/{max_regeneration}
+              </ProgressBar>
+              <Section>
```
```suggestion

```

jlsnow301 commented on tgui/packages/tgui/interfaces/ModularShieldGen.tsx:
```diff
@@ -0,0 +1,126 @@
+  return (
+    <Window title="Modular Shield Generator" width={900} height={225}>
+      <Window.Content>
+        <Stack fill>
+          <Stack.Item grow={2}>
```
if you no longer want this stacked horizontally vs settings, remove these tags and the latter stack.item as well. just giving an example how to use flex

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag committed:
Update tgui/packages/tgui/interfaces/ModularShieldGen.tsx

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

moocowswag commented:
![image](https://github.com/tgstation/tgstation/assets/62126254/70d5ce30-af29-4948-9e83-9f031dcffcc6)

moocowswag committed:
widenuminput

moocowswag committed:
ui

moocowswag committed:
uiadjust

moocowswag commented:
![image](https://github.com/tgstation/tgstation/assets/62126254/7d540878-1974-40b8-8c7f-d4417959ad9a)

moocowswag committed:
Merge branch 'master' into ShieldGeneratorModular

Loafin34 commented:
moocow, if you could make directional shields with more health, that'd be really fun. I'd love to barricade my fort with several arrays of shield walls while springing traps on the bad guys.

moocowswag commented:
> moocow, if you could make directional shields with more health, that'd be really fun. I'd love to barricade my fort with several arrays of shield walls while springing traps on the bad guys.

I plan on having more ways to generate shields in the future, bubble/circle will just be the default configuration.

All of the future configurations will still be reliant on the radius stat to actually reach the tiles you want to fill in though.

moocowswag commented:
This pr is pretty much complete unless san or someone else requests anything else. I would like a test merge before a full merge first just to see if there is anything i missed though.

moocowswag committed:
lazylistdoneandtested

moocowswag commented:
forgot san wanted to do lazy lists, its done now and fully tested.

moocowswag committed:
Merge branch 'master' into ShieldGeneratorModular

Loafin34 commented:
I think it'd be a little bit better if you added some descriptions to how you manipulate it through tools. Something along the lines of "It's anchored to one angle." or "The connectors seem to require pulsing before they connect." good luck, otherwise.

moocowswag commented:
> I think it'd be a little bit better if you added some descriptions to how you manipulate it through tools. Something along the lines of "It's anchored to one angle." or "The connectors seem to require pulsing before they connect." good luck, otherwise.

fiiiinnne

moocowswag committed:
tooldescriptions

moocowswag commented:
> I think it'd be a little bit better if you added some descriptions to how you manipulate it through tools. Something along the lines of "It's anchored to one angle." or "The connectors seem to require pulsing before they connect." good luck, otherwise.

done

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
 #undef ACTIVE_HASFIELDS
+
+//Modular Shield Generator Start
+/obj/machinery/modular_shield_generator
```
Please put this in its own file

moocowswag replied:
ill just name it modular_shield_gen.dm if thats not too long

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
+
+/obj/machinery/modular_shield_generator/attackby(obj/item/W, mob/user, params)
+
+	if(default_deconstruction_screwdriver(user,"gen_[!(machine_stat & NOPOWER) ? "[recovering ? "recovering_" : "ready_"]" : "no_power_"]open",
```
Screwdriver act instead

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
+		"gen_[!(machine_stat & NOPOWER) ? "[recovering ? "recovering_" : "ready_"]" : "no_power_"]closed",  W))
+		return
+
+	if(default_deconstruction_crowbar(W) && !(active) && !(recovering))
```
Crowbar act instead

moocowswag replied:
done

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
+///The general code used for machines that want to connect to the network
+/obj/machinery/modular_shield/module
+
+	name = "Modular Shield Debugger" //Filler name and sprite for testing
```
Are you gonna change the name for this?

moocowswag replied:
You arnt meant to be able to obtain this in game, its a basetype incase anyone wants to make their own subtype but didnt give it a name or sprite.

Do you want a different name for this?

JohnFulpWillard replied:
I read "for testing" and assumed you wanted it changed before merging. It's fine then

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
+
+/obj/machinery/modular_shield/module/attackby(obj/item/I, mob/user, params)
+
+	if(I.tool_behaviour == TOOL_SCREWDRIVER)
```
Screwdriver act

moocowswag replied:
done

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
+
+	//rather than automatically checking for connections its probably alot less
+	//expensive to just make the players manually multi tool sync each part
+	if(I.tool_behaviour == TOOL_MULTITOOL)
```
Multitool act (unless you want wirecutters to do this too, like parent's attackby

moocowswag replied:
done, multitool only

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
+		try_connect(user)
+		return TRUE
+
+	if(default_change_direction_wrench(user, I))
```
Wrench act

moocowswag replied:
done

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
+		connected_turf = get_step(loc, dir)
+		return TRUE
+
+	if(default_deconstruction_crowbar(I))
```
Crowbar act

moocowswag replied:
done

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
+		shield_generator.calculate_boost()
+		return
+
+	connected_node	= (locate(/obj/machinery/modular_shield/module/node) in connected_turf)
```
```suggestion
	connected_node = (locate(/obj/machinery/modular_shield/module/node) in connected_turf)
```

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
+			return
+		balloon_alert(user, "connected to node")
+		return
+	balloon_alert(user, "failed to find connection!")
```
```suggestion
	balloon_alert(user, "no connection!")
```

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
+		shield_generator = connected_node.shield_generator
+		if(shield_generator)
+			LAZYOR(shield_generator.connected_modules, (src))
+			balloon_alert(user, "connected to generator through node")
```
```suggestion
			balloon_alert(user, "connected to generator")
```

JohnFulpWillard commented on code/game/machinery/shieldgen.dm:
```diff
@@ -553,3 +553,668 @@
+	density = FALSE
+	alpha = 100
+	resistance_flags = INDESTRUCTIBLE //the shield itself is indestructible or atleast should be
+	//our parent
```
Parent what?

moocowswag replied:
changed to ///the shield generator sustaining us

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

moocowswag committed:
Update code/game/machinery/shieldgen.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

moocowswag committed:
tool_acts

moocowswag committed:
shieldgentool_acts

moocowswag committed:
copything

moocowswag committed:
transfertonewfile

moocowswag committed:
shortername

moocowswag committed:
whyisthisathing

moocowswag committed:
Merge branch 'master' into ShieldGeneratorModular

moocowswag committed:
commentwasinwrongspot

JohnFulpWillard commented on code/__HELPERS/maths.dm:
```diff
@@ -80,6 +80,41 @@
+ * Uses [Jesko`s method to the midpoint circle Algorithm](https://en.wikipedia.org/wiki/Midpoint_circle_algorithm).
+ */
+/proc/get_perimeter(atom/center, radius)
+	if(radius < 0)
```
This will not early return if radius is 0, is that normal?

moocowswag replied:
youre right i should exclude 0

JohnFulpWillard commented on code/game/machinery/modular_shield.dm:
```diff
@@ -0,0 +1,681 @@
+	tool.play_tool_sound(src, 50)
+	update_icon_state()
+	if(panel_open)
+		to_chat(user, span_notice("You open the maintenance hatch of [src]."))
+		return TRUE
+	to_chat(user, span_notice("You close the maintenance hatch of [src]."))
```
Should these be balloon alert?

moocowswag replied:
I tried to match the normal messages you get when screwdrivering other machines, are we moving away from that?

moocowswag replied:
done

JohnFulpWillard commented on code/game/machinery/modular_shield.dm:
```diff
@@ -0,0 +1,681 @@
+		LAZYOR(shield_generator.connected_modules, (src))
+		balloon_alert(user, "connected to generator")
+		update_icon_state()
+		if(istype(src, /obj/machinery/modular_shield/module/node))
+			var/obj/machinery/modular_shield/module/node/connected_node = src
+			connected_node.connect_connected_through_us()
```
We should never be checking src for type, you should move this to its subtype instead.

moocowswag replied:
done

JohnFulpWillard commented on code/game/machinery/modular_shield.dm:
```diff
@@ -0,0 +1,681 @@
+			LAZYOR(shield_generator.connected_modules, (src))
+			balloon_alert(user, "connected to generator")
+			update_icon_state()
+			if(istype(src, /obj/machinery/modular_shield/module/node))
+				var/obj/machinery/modular_shield/module/node/connected_node = src
+				connected_node.connect_connected_through_us()
```
Ditto

moocowswag replied:
done

JohnFulpWillard commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +167,53 @@
+	greyscale_colors = CIRCUIT_COLOR_ENGINEERING
+	build_path = /obj/machinery/modular_shield/module/node
+	req_components = list(
+	/obj/item/stack/cable_coil = 15,
+	/obj/item/stack/sheet/plasteel = 2,
```
Add another tab here please

moocowswag replied:
done, and how the fuck did that happen

JohnFulpWillard commented on code/game/objects/items/circuitboards/machines/machine_circuitboards.dm:
```diff
@@ -159,6 +167,53 @@
+	greyscale_colors = CIRCUIT_COLOR_ENGINEERING
+	build_path = /obj/machinery/modular_shield/module/well
+	req_components = list(
+			/datum/stock_part/capacitor = 3,
+			/obj/item/stack/sheet/plasteel = 2,
```
```suggestion
		/datum/stock_part/capacitor = 3,
		/obj/item/stack/sheet/plasteel = 2,
```

moocowswag replied:
tf, done

moocowswag committed:
exlude radius <1

moocowswag committed:
movedconnecttonode

moocowswag committed:
indents

moocowswag committed:
balloons
